elasticsearch/src/main/java/com/amazon/opendistroforelasticsearch/sql/elasticsearch/storage/script/filter/ExpressionFilterScript.java

```diff
@@ -0,0 +1,169 @@
+ *    on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *    express or implied. See the License for the specific language governing
+ *    permissions and limitations under the License.
+ *
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.elasticsearch.storage.script.filter;
+
+import static com.amazon.opendistroforelasticsearch.sql.data.type.ExprCoreType.FLOAT;
+import static com.amazon.opendistroforelasticsearch.sql.data.type.ExprCoreType.INTEGER;
+import static com.amazon.opendistroforelasticsearch.sql.elasticsearch.data.type.ElasticsearchDataType.ES_TEXT_KEYWORD;
+import static java.util.stream.Collectors.toMap;
+
+import com.amazon.opendistroforelasticsearch.sql.data.model.ExprBooleanValue;
+import com.amazon.opendistroforelasticsearch.sql.data.model.ExprTupleValue;
+import com.amazon.opendistroforelasticsearch.sql.data.model.ExprValue;
+import com.amazon.opendistroforelasticsearch.sql.data.type.ExprCoreType;
+import com.amazon.opendistroforelasticsearch.sql.data.type.ExprType;
+import com.amazon.opendistroforelasticsearch.sql.elasticsearch.data.value.ElasticsearchExprValueFactory;
+import com.amazon.opendistroforelasticsearch.sql.expression.Expression;
+import com.amazon.opendistroforelasticsearch.sql.expression.ExpressionNodeVisitor;
+import com.amazon.opendistroforelasticsearch.sql.expression.ReferenceExpression;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.time.chrono.ChronoZonedDateTime;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import lombok.EqualsAndHashCode;
+import org.apache.lucene.index.LeafReaderContext;
+import org.elasticsearch.SpecialPermission;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.script.FilterScript;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+/**
+ * Expression script executor that executes the expression on each document
+ * and determine if the document is supposed to be filtered out or not.
+ */
+@EqualsAndHashCode(callSuper = false)
+class ExpressionFilterScript extends FilterScript {
+
+  /**
+   * Expression to execute.
+   */
+  private final Expression expression;
+
+  public ExpressionFilterScript(Expression expression,
+                                SearchLookup lookup,
+                                LeafReaderContext context,
+                                Map<String, Object> params) {
+    super(params, lookup, context);
+    this.expression = expression;
+  }
+
+  @Override
+  public boolean execute() {
+    // Check we ourselves are not being called by unprivileged code.
+    SpecialPermission.check();
+
+    return AccessController.doPrivileged((PrivilegedAction<Boolean>) () -> {
+      Set<ReferenceExpression> fields = extractFields(expression);
+      ElasticsearchExprValueFactory valueFactory = buildValueFactory(fields);
+      Map<String, ExprValue> valueEnv = buildValueEnv(fields, valueFactory);
```

Comment:
NP, Could we impl BindingTuple and encapsulate these logic in that to achive the same concept of valueEnv.

Reply:
Sure, will add a new `Environment` class to encapsulate the map data structure. Thanks!